packages/pnpm-sync-lib/src/pnpmSyncPrepare.ts

```diff
@@ -260,47 +267,19 @@ export async function pnpmSyncPrepareAsync(options: IPnpmSyncPrepareOptions): Pr
   });
 }
 
-// process dependencies and devDependencies to generate injectedDependencyToFilePath
-function getInjectedDependencyToVersion(pnpmLockfile: ILockfile | undefined): Map<string, Set<string>> {
-  const injectedDependencyToVersion: Map<string, Set<string>> = new Map();
-  for (const importerKey in pnpmLockfile?.importers) {
-    if (!pnpmLockfile?.importers[importerKey]?.dependenciesMeta) {
-      continue;
-    }
-    const dependenciesMeta = pnpmLockfile?.importers[importerKey]?.dependenciesMeta;
-
-    for (const dependency in dependenciesMeta) {
-      if (dependenciesMeta[dependency]?.injected) {
-        if (!injectedDependencyToVersion.has(dependency)) {
-          injectedDependencyToVersion.set(dependency, new Set());
-        }
-      }
-    }
-
-    // based on https://pnpm.io/package_json#dependenciesmeta
-    // the injected dependencies could available inside dependencies, optionalDependencies, and devDependencies.
-    processDependencies(pnpmLockfile?.importers[importerKey]?.dependencies, injectedDependencyToVersion);
-    processDependencies(pnpmLockfile?.importers[importerKey]?.devDependencies, injectedDependencyToVersion);
-    processDependencies(
-      pnpmLockfile?.importers[importerKey]?.optionalDependencies,
-      injectedDependencyToVersion
-    );
-  }
-  return injectedDependencyToVersion;
-}
-function processDependencies(
+function getInjectedDependencyToVersion(
   dependencies: Record<string, IVersionSpecifier> | undefined,
   injectedDependencyToVersion: Map<string, Set<string>>
 ): void {
   if (dependencies) {
     for (const [dependency, specifier] of Object.entries(dependencies)) {
-      if (injectedDependencyToVersion.has(dependency)) {
-        const specifierToUse: string = typeof specifier === 'string' ? specifier : specifier.version;
-
-        // the injected dependency should always start with file protocol
-        if (specifierToUse.startsWith('file:')) {
-          injectedDependencyToVersion.get(dependency)?.add(specifierToUse);
+      const specifierToUse: string = typeof specifier === 'string' ? specifier : specifier.version;
+      // the injected dependency should always start with file protocol
+      if (specifierToUse.startsWith('file:')) {
```

Comment:
Consider a **package.json** like this:

```js
{
  "name": "example",
  "version": "1.0.0",
  "dependencies": {
    "colors": "file:./tarball/colors-1.4.0.tgz",
    "thing": "file:./inject/"
  }
}
```

The resulting lockfile is like this:

**pnpm-lock.yaml**
```yaml
lockfileVersion: '6.0'

settings:
  autoInstallPeers: true
  excludeLinksFromLockfile: false

dependencies:
  colors:
    specifier: file:./tarball/colors-1.4.0.tgz
    version: file:tarball/colors-1.4.0.tgz
  thing:
    specifier: file:./inject/
    version: file:inject

packages:

  file:inject:
    resolution: {directory: inject, type: directory}
    name: thing
    dev: false

  file:tarball/colors-1.4.0.tgz:
    resolution: {integrity: sha512-a+UqTh4kgZg/SlGvfbzDHpgRu7AAQOmmqRHJnxhRZICKFUT91brVhNNt58CMWU9PsBbv3PDCZUHbVxuDiH2mtA==, tarball: file:tarball/colors-1.4.0.tgz}
    name: colors
    version: 1.4.0
    engines: {node: '>=0.1.90'}
    dev: false
```

`colors` should be injected but `thing` should not.

Reply:
Why `thing` should not? We also need to copy `thing` to the `.pnpm` folder and, whenever `thing` get updates, need to sync to `.pnpm` folder as well?

Reply:
@g-chao sorry, my comment was backwards.  I should have said:

`thing` should be injected but `colors` should NOT.

Reply:
Ok, the solution I can think to exclude these tarball file (.tar, .tar.gz, or .tgz, based on https://pnpm.io/cli/add). 

A more robust solution is, on Rush side, when process `alwaysInjectDependenciesFromOtherSubspaces`, write `injected:true` to its package.json. The drawback is we need to modify package.json file

Reply:
@octogonz  I think we will need to go with the robust solution to prevent potential trouble in the future.  That is the `injected:true` in the package.json is the source of truth to tell if a dependency is injected or not. 

So, for this issue, we need to fix on Rush side, https://github.com/microsoft/rushstack/pull/4724